### ISSUE TITLE
DecisionModel ignores head_checkpointing enabled by the fine-tuning notebook

### ISSUE THREAD

**Issue by zhangxinping666** [NandhaKishorM/laya#148]
The official fine-tuning notebook sets `model.head_checkpointing = True`, but the decision head does not use that setting.

At main `573e5b6`, [`DecisionModel.__init__`](https://github.com/NandhaKishorM/laya/blob/573e5b62696ba441230cd6be71d593331b5d23af/laya/common.py#L103) initializes the flag, while the head-layer loop in `forward` always calls each layer directly. The [typed-decisions notebook](https://github.com/NandhaKishorM/laya/blob/573e5b62696ba441230cd6be71d593331b5d23af/notebooks/laya_finetune_typed_decisions_2xT4_kaggle.ipynb) enables the flag after enabling encoder gradient checkpointing. The encoder setting does not cover these separate head layers.

### Offline reproduction

Run from a source checkout with torch and transformers installed. This uses randomly initialized local weights and does not download a checkpoint.

```python
import torch
from transformers import BertConfig, BertModel
from laya.common import DecisionModel

torch.set_num_threads(1)
cfg = BertConfig(vocab_size=32, hidden_size=16, num_hidden_layers=1,
                 num_attention_heads=1, intermediate_size=32)
model = DecisionModel(BertModel(cfg), head_layers=1).train()
calls = []
hook = model.head.layers[0].register_forward_pre_hook(
    lambda module, args: calls.append(1)
)
ids = torch.randint(0, 32, (1, 8))
positions = torch.tensor([[2, 5]])
for enabled in (False, True):
    model.zero_grad(set_to_none=True)
    model.head_checkpointing = enabled
    calls.clear()
    logits, act = model(
        ids, torch.ones_like(ids), positions,
        torch.ones_like(positions, dtype=torch.bool),
        torch.zeros(1, dtype=torch.long),
    )
    (logits.square().mean() + act.square().mean()).backward()
    print(f"head_checkpointing={enabled}: {len(calls)} layer calls")
hook.remove()
```

Actual on main: **1 call with the flag off and 1 call with it on**. The enabled path does not recompute the head during backward.

Expected: when the flag is enabled during gradient-enabled training, head activations should be checkpointed and recomputed for backward. Evaluation, no-grad execution and the default disabled path should keep their current behavior.

### Impact and proposed fix

The notebook's requested memory-saving behavior is silently absent for the decision head. This is an activation-retention issue; I have not reproduced an OOM in the published T4 training workflow.

A focused fix can use PyTorch's non-reentrant checkpointing in the existing head-layer loop, preserving dropout RNG state. I have reproduced the missing recomputation on CPU and prepared regression coverage for outputs, gradients, optimizer updates, frozen encoder inputs and encoder/head checkpointing together.

Related: #53 also sets this flag in its proposed training entry point, but does not change the shared `DecisionModel.forward` implementation. This fix would support that caller as well.

Reproduced on Windows, Python 3.11.13, torch 2.14.0+cpu and transformers 5.17.0.

**Comment by NandhaKishorM** [comment]
Confirmed on `main`: `DecisionModel.__init__` sets `head_checkpointing = False`, the fine-tuning notebook sets it to `True`, and nothing in `forward` reads it. Your repro makes that unambiguous, and thanks for being clear that you haven't reproduced an OOM, only the missing recomputation.

There are two PRs, #149 (yours) and #160. I'll review yours first since it carries the gradient and optimizer-update regression coverage, which is what makes a checkpointing change safe to merge.

**Comment by NandhaKishorM** [comment]
An update: #149 (yours) is the fix I'm taking. It carries the gradient and optimizer-update regression coverage. After today's merges it only conflicts in `README.md`, and it lands as soon as it's rebased.
